File: algorithms/js/51.n皇后.js
/*
 * @lc app=leetcode.cn id=51 lang=javascript
 *
 * [51] N皇后
 */

// @lc code=start
function isValid(x, y, nums) {
  // left
  for (let i = x; i > -1; i--) {
    if (nums[i][y] === "Q") return false;
  }
  // top
  for (let i = y; i > -1; i--) {
    if (nums[x][i] === "Q") return false;
  }
  // top-left
  for (let i = x, j = y; i > -1 && j > -1; i--, j--) {
    if (nums[i][j] === "Q") return false;
  }
  // top-right
  for (let i = x, j = y; i > -1 && j < nums.length; i--, j++) {
    if (nums[i][j] === "Q") return false;
  }
  return true;
}
/**
 * @param {number} row
 * @param {number} n
 * @param {string[]} curAns
 * @param {string[][]} ans
 */
function dfs(row, n, curAns, ans) {
  if (row == n) {
    ans.push(curAns.map((charArr) => charArr.join("")));
    return;
  }
  for (let i = 0; i < n; i++) {
    if (isValid(row, i, curAns)) {
      curAns[row][i] = "Q";
      dfs(row + 1, n, [...curAns], ans);
      curAns[row][i] = ".";
    }
  }
}
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function(n) {
  const ans = [];
  if (n === 0) return ans;
  const tmp = Array.from({ length: n }, () => "".padStart(n, ".").split(""));
  dfs(0, n, tmp, ans);
  return ans;
};
// @lc code=end

File: algorithms/cpp/51.n-皇后.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;
/*
 * @lc app=leetcode.cn id=51 lang=cpp
 *
 * [51] N皇后
 * 皇后能攻击已当前位置为原点的 0°, 45°, 90°, 135° 的直线上的旗子.
 */

// @lc code=start
class Solution {
 private:
  bool isValid(const int& row, const int& col, const vector<string>& board) {
    // top
    for (int i = row - 1; i > -1; i--) {
      if (board[i][col] == 'Q') return false;
    }

    // top-left
    for (int i = row - 1, j = col - 1; i > -1 && j > -1; i--, j--) {
      if (board[i][j] == 'Q') return false;
    }

    // top-right
    for (int i = row - 1, j = col + 1; i > -1 && j < board[i].size();
         i--, j++) {
      if (board[i][j] == 'Q') return false;
    }

    return true;
  }
  void dfs(int row, int n, vector<string>& board,
           vector<vector<string> >& ans) {
    //    terminator
    if (row == n) {
      ans.push_back(board);
      return;
    }

    // process current logic
    // ...

    for (int col = 0; col < n; col++) {
      if (isValid(row, col, board)) {
        // drill down
        board[row][col] = 'Q';
        dfs(row + 1, n, board, ans);
        // restore state
        board[row][col] = '.';
      }
    }
  }

 public:
  vector<vector<string> > solveNQueens(int n) {
    vector<vector<string> > ans;
    // generate board
    vector<string> board(n, string(n, '.'));
    dfs(0, n, board, ans);
    return ans;
  }
};
// @lc code=end
